### PR TITLE
Add UNION to iterator; closes #4914

### LINKED ISSUE #4914: Add UNION support to iterator
A few queries currently uses UNION, this is NOT possible to really migrate them to the iterator until it support has been added.

### PR DESCRIPTION
**Pull request by trasher**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #4914 

Right now, I've implemented the capability and the tests; that's what I want to validate right now. `DocumentItem` changes are WIP and can be ignored right now :)

### CODE REVIEW

File: inc/dbmysqliterator.class.php [hunk 1]
```diff
@@ -289,7 +289,11 @@ function buildQuery ($table, $crit = "", $log = false) {
                trigger_error("Missing table name", E_USER_ERROR);
             }
          } else if ($table) {
-            $table = DBmysql::quoteName($table);
+            if ($table instanceof \QueryUnion) {
```
Comment (cedric-anne) [inline inc/dbmysqliterator.class.php]: The leading `\` is not mandatory.

File: inc/dbmysqliterator.class.php [hunk 2]
```diff
@@ -396,7 +400,11 @@ public function handleLimits($limit, $offset = null) {
     */
    private function handleFields($t, $f) {
       if (is_numeric($t)) {
-         return DBmysql::quoteName($f);
+         if ($f instanceof \QueryUnion) {
```
Comment (cedric-anne) [inline inc/dbmysqliterator.class.php]: The leading `\` is not mandatory.

File: inc/document_item.class.php [hunk 3]
```diff
@@ -728,31 +728,64 @@ static function showListForItem(CommonDBTM $item, $withtemplate = 0, $options =
...
       } else {
          // Anonymous access from FAQ
-         $query .= " AND `glpi_documents`.`entities_id`= '0' ";
+         $criteria['WHERE']['glpi_documents.entities_id'] = 0;
       }
 
       // Document : search links in both order using union
       if ($item->getType() == 'Document') {
-         $query .= "UNION
+         $doc_criteria = $criteria;
```
Comment (cedric-anne) [inline inc/document_item.class.php]: JOIN criteria differs on `glpi_documents` (`glpi_documents_items.documents_id` VS `glpi_documents_items.items_id`) https://github.com/glpi-project/glpi/pull/4925/files#diff-f6061ea857b63b7935db5eaddac48c64R797

File: inc/document_item.class.php [hunk 4]
```diff
@@ -728,31 +728,64 @@ static function showListForItem(CommonDBTM $item, $withtemplate = 0, $options =
...
+         unset($doc_criteria['WHERE']['glpi_documents_items.items_id']);
+         unset($doc_criteria['WHERE']['glpi_documents_items.itemtype']);
+
+         $criteria['WHERE'] = $criteria['WHERE'] + [
```
Comment (cedric-anne) [inline inc/document_item.class.php]:
```suggestion
         $doc_criteria['WHERE'] = $doc_criteria['WHERE'] + [
```

File: inc/document_item.class.php [hunk 5]
```diff
@@ -728,31 +728,64 @@ static function showListForItem(CommonDBTM $item, $withtemplate = 0, $options =
...
+            'glpi_documents_items.documents_id' => $item->getID(),
+            'glpi_documents_items.itemtype'     => $item->getType()
```
Comment (cedric-anne) [inline inc/document_item.class.php]: Not sure this criteria is OK. If we want to match on `documents_id` field, then we should not take care of `itemtype` (or maybe I do not understant which result we want)..

File: inc/document_item.class.php [hunk 6]
```diff
@@ -766,17 +799,19 @@ static function showListForItem(CommonDBTM $item, $withtemplate = 0, $options =
...
             $query .= getEntitiesRestrictRequest(" AND", "glpi_documents", '', '', true);
          } else {
             // Anonymous access from FAQ
             $query .= " AND `glpi_documents`.`entities_id`='0' ";
-         }
+         }*/
       }
-      $query .= " ORDER BY $sort $order";
 
       $result = $DB->query($query);
```
Comment (cedric-anne) [inline inc/document_item.class.php]: `$query` does not exist anymore. You have to use a `QueryUnion` object here.

File: inc/queryunion.class.php [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+   private $queries = [];
+   private $alias;
+   private $dups = true;
+
+   /**
+    * Create a sub query
+    *
+    * @param array   $queries An array of queries to union. Either SubQuery objects
+    *                         or an array of criteria to build them.
+    * @param boolean $dups    Include duplicates. Turning off may has
```
Comment (cedric-anne) [inline inc/queryunion.class.php]: I would have do `$distinct = false`, as `distinct` is a more common name for querying features.

File: tests/units/DBmysqlIterator.php [hunk 8]
```diff
@@ -717,4 +717,42 @@ function() use($crit) {
...
+      $union_crit = [
+         ['FROM' => 'table1'],
+         ['FROM' => 'table2']
+      ];
+      $union = new \QueryUnion($union_crit);
+      $raw_query = 'SELECT * FROM (SELECT * FROM `table1` UNION ALL SELECT * FROM `table2`)';
+      $it = $this->it->execute(['FROM' => $union]);
+      $this->string($it->getSql())->isIdenticalTo($raw_query);
+
+      $raw_query = 'SELECT (SELECT * FROM `table1` UNION ALL SELECT * FROM `table2`) FROM `table`';
```
Comment (cedric-anne) [inline tests/units/DBmysqlIterator.php]: It does not seems to be a valid query. If passing `QueryUnion` to `SELECT` is accepted, we should find a real working case.

File: inc/dbmysqliterator.class.php (hunk 1 above)
Comment (trasher) [reply inc/dbmysqliterator.class.php]: I know

File: inc/document_item.class.php (hunk 3 above)
Comment (trasher) [reply inc/document_item.class.php]: Indeed, but this is not finished yet :)

File: inc/document_item.class.php (hunk 5 above)
Comment (trasher) [reply inc/document_item.class.php]: This was previously:
```
                    WHERE `glpi_documents_items`.`documents_id` = '".$item->getID()."'
                          AND `glpi_documents_items`.`itemtype` = '".$item->getType()."' "; 
```
So for now I've just migrated.

But indeed I agrre it should be on ether `document_id` or `itemtype + items_id`

File: inc/queryunion.class.php (hunk 7 above)
Comment (trasher) [reply inc/queryunion.class.php]: @orthagh ?

File: inc/document_item.class.php [hunk 9]
```diff
@@ -728,67 +728,74 @@ static function showListForItem(CommonDBTM $item, $withtemplate = 0, $options =
...
+         $doc_criteria['WHERE'] = $doc_criteria['WHERE'] + [
+            'glpi_documents_items.documents_id' => $item->getID()
+         ];
+         $criteria = new \QueryUnion([$criteria, $doc_criteria]);
       }
-      $query .= " ORDER BY $sort $order";
 
-      $result = $DB->query($query);
-      $number = $DB->numrows($result);
+      $iterator = $DB->request();
```
Comment (cedric-anne) [inline inc/document_item.class.php]: Something is missing.

```suggestion
      $iterator = $DB->request(['FROM' => $criteria]);
```

File: inc/document_item.class.php [hunk 10]
```diff
@@ -728,67 +728,74 @@ static function showListForItem(CommonDBTM $item, $withtemplate = 0, $options =
...
-            $query .= " AND `glpi_documents`.`entities_id`='0' ";
-         }
+         $doc_criteria = $criteria;
+         unset($doc_criteria['WHERE']['glpi_documents_items.items_id']);
+         unset($doc_criteria['WHERE']['glpi_documents_items.itemtype']);
+
+         $doc_criteria['WHERE'] = $doc_criteria['WHERE'] + [
+            'glpi_documents_items.documents_id' => $item->getID()
+         ];
+         $criteria = new \QueryUnion([$criteria, $doc_criteria]);
```
Comment (cedric-anne) [inline inc/document_item.class.php]:
```suggestion
         $criteria = ['FROM' => new \QueryUnion([$criteria, $doc_criteria])];
```

File: tests/units/DBmysqlIterator.php (hunk 8 above)
Comment (trasher) [reply tests/units/DBmysqlIterator.php]: It was a mistake. I've replaced the test with a complex real example of what is expected to work.

File: inc/document_item.class.php [hunk 11]
```diff
@@ -728,67 +728,74 @@ static function showListForItem(CommonDBTM $item, $withtemplate = 0, $options =
...
-            $query .= " AND `glpi_documents`.`entities_id`='0' ";
-         }
+         $doc_criteria = $criteria;
+         unset($doc_criteria['WHERE']['glpi_documents_items.items_id']);
+         unset($doc_criteria['WHERE']['glpi_documents_items.itemtype']);
+
+         $doc_criteria['WHERE'] = $doc_criteria['WHERE'] + [
+            'glpi_documents_items.documents_id' => $item->getID()
+         ];
+         $criteria['FROM'] = new \QueryUnion([$criteria, $doc_criteria]);
```
Comment (cedric-anne) [inline inc/document_item.class.php]: You have to redifine the whole criteria object, to remove 'SELECT', 'WHERE', ... keys.
```suggestion
         $criteria = ['FROM' => new \QueryUnion([$criteria, $doc_criteria])];
```

**Pull request merged**